File: tests/test_emulator_queue2_attacks.py
"""S09-S12 native canvas acceptance against fresh offline Wasm."""
import unittest
from emulator_browser import objects,click_text,click_object
import test_emulator_portal_demo as portal_helpers

class Queue2AttacksBrowser(unittest.TestCase):
    setUp=portal_helpers.PortalDemoBrowser.setUp
    load=portal_helpers.PortalDemoBrowser.load
    ready=portal_helpers.PortalDemoBrowser.ready
    advance=portal_helpers.PortalDemoBrowser.advance
    clean=portal_helpers.PortalDemoBrowser.clean
    scan=portal_helpers.PortalDemoBrowser.scan
    def begin(self,id,rotation=0):
        self.load(rotation)
        self.page.locator('#queue2-story').select_option(id)
        self.page.locator('#guide-queue2').click()
        self.page.evaluate("async id=>{const {queue2Attacks}=await import('./queue2-attacks.mjs');globalThis.qaStory=queue2Attacks.find(e=>e.id===id).create().story;}",id)
    def stage(self,n):
        self.page.wait_for_function('(n)=>document.querySelector("#guide-title").textContent===(n===qaStory.steps.length?qaStory.completionTitle:qaStory.steps[n].title)',arg=n,timeout=7000)
    def done(self):
        self.page.wait_for_function('document.querySelector("#guide-title").textContent===qaStory.completionTitle',timeout=7000)
        self.assertIn('complete',self.page.locator('#guide-progress').inner_text());self.clean()
    def tap(self,fragment):
        click_object(self.page,next(o for o in objects(self.page) if fragment in o['binding']))
    def test_scan_actions_success_cancel_and_rotations(self):
        for rotation in range(4):
            for id,label,stop in [('deauth','Deauth','scan_deauth_popup_close_cb'),('sae','SAE Overflow','show_sae_popup.stop_btn'),('handshake','Handshake','show_handshaker_popup')]:
                for cancel in (False,True):
                    with self.subTest(rotation=rotation,id=id,cancel=cancel):
                        self.begin(id+('-cancel' if cancel else ''),rotation);self.scan();self.stage(1)
                        click_text(self.page,label);self.stage(2)
                        if cancel:self.page.locator('#cancel').click();self.stage(3)
                        else:self.stage(3)
                        if id=='handshake':self.tap('show_handshaker_popup.ctx--handshaker_stop_btn')
                        elif id=='deauth':self.tap('show_scan_deauth_popup.stop_btn')
                        else:self.tap(stop)
                        self.done()
    def test_global_confirm_no_yes_stop_and_rotations(self):
        for rotation in range(4):
            for id,label,kind in [('blackout','Blackout','blackout'),('global-handshaker','Handshaker','global_handshaker'),('snifferdog','SnifferDog','snifferdog')]:
                for cancel in (False,True):
                    with self.subTest(rotation=rotation,id=id,cancel=cancel):
                        self.begin(id+('-cancel' if cancel else ''),rotation);click_text(self.page,'Global WiFi\nAttacks');click_text(self.page,label);self.stage(1)
                        click_text(self.page,'No');self.stage(2)
                        if cancel:self.tap('show_global_attacks_page.back_btn')
                        else:
                            click_text(self.page,label);click_text(self.page,'Yes');self.stage(3);self.stage(4)
                            self.tap('show_'+kind+'_active_popup.stop_btn')
                        self.done()
    def test_karma_menu_and_rotations(self):
        for rotation in range(4):
            for id in ['karma-menu']:
                with self.subTest(rotation=rotation,id=id):
                    self.begin(id,rotation)
                    self.stage(1);click_text(self.page,'Karma');self.stage(2);click_text(self.page,'Start Sniffer');self.stage(3);self.stage(4)
                    click_text(self.page,'Stop Sniffer');self.page.wait_for_timeout(100);self.tap('emu.phase45.karma.probe.0.0');self.stage(5);click_text(self.page,'Start Karma');self.stage(6);self.stage(7)
                    click_text(self.page,'STOP');self.stage(8);self.tap('emu.phase45.show_karma_page.back_btn');self.done()
    def test_beacon_list_start_empty_and_rotations(self):
        for rotation in range(4):
            for empty in (False,True):
                with self.subTest(rotation=rotation,empty=empty):
                    self.begin('beacon-empty' if empty else 'beacon',rotation);click_text(self.page,'Global WiFi\nAttacks');click_text(self.page,'Beacon Spam');click_text(self.page,'List SSIDs');self.stage(1)
                    if empty:
                        for _ in range(2):self.tap('beacon_spam_rebuild_ssid_grid.delete_btn')
                        self.stage(2);self.tap('show_beacon_ssids_page.back_btn');click_text(self.page,'Start Spam');self.stage(3);self.tap('show_beacon_spam_page.back_btn')
                    else:
                        self.tap('show_beacon_ssids_page.back_btn');click_text(self.page,'Start Spam');self.stage(2);self.stage(3);self.tap('emu.phase45.beacon_spam_start_cb.beacon_spam_active_close_cb')
                    self.done()
    def test_handshake_wrong_target_and_busy_refusal(self):
        self.begin('handshake');self.scan('02:20:77:00:00:02')
        click_text(self.page,'Handshake');self.advance(3000)
        self.assertEqual(self.page.locator('#guide-title').inner_text(),'Select NEON-BAZAAR')
        job=self.page.evaluate('emulator.device.device.job(emulator.device.snapshot(0).active)')
        self.assertEqual(job['options']['networkIds'],['net-2'])
        self.assertIn('02:20:77:00:00:02',str(self.page.evaluate('emulator.device.snapshot(0).networks')))
        self.tap('show_handshaker_popup.ctx--handshaker_stop_btn');self.clean()
        self.begin('handshake');self.scan();self.stage(1)
        job=self.page.evaluate("emulator.device.device.attackStart('grove','karma')")
        click_text(self.page,'Handshake');self.advance(500)
        self.assertEqual(self.page.evaluate('emulator.device.snapshot(0).active'),job)
        self.assertEqual(self.page.locator('#guide-title').inner_text(),'Start Handshake')
        self.assertIn('busy',' '.join(o['text'] for o in objects(self.page)))
        self.tap('show_handshaker_popup.ctx--handshaker_stop_btn');self.page.locator('#cancel').click();self.clean()
    def test_handshake_no_result_does_not_claim_capture(self):
        self.begin('handshake');self.scan('02:20:77:00:00:04');click_text(self.page,'Handshake');self.advance(4000)
        text=' '.join(o['text'] for o in objects(self.page))
        self.assertIn('No synthetic handshake available',text);self.assertNotIn('Handshake captured:',text)
        self.assertEqual(self.page.locator('#guide-title').inner_text(),'Select NEON-BAZAAR')
        self.tap('show_handshaker_popup.ctx--handshaker_stop_btn');self.clean()
    def test_restart_wrong_module_disconnect_and_busy(self):
        self.begin('deauth');self.scan();self.stage(1);click_text(self.page,'Deauth');self.stage(2)
        self.page.locator('#guide-restart').click();self.page.wait_for_timeout(300)
        self.assertNotIn('complete',self.page.locator('#guide-progress').inner_text())
        click_text(self.page,'MBUS');self.page.wait_for_timeout(200);self.assertNotIn('complete',self.page.locator('#guide-progress').inner_text())
        click_text(self.page,'GROVE');self.page.evaluate("emulator.device.device.setModule('grove',{connected:false})");self.advance(300)
        self.assertNotIn('complete',self.page.locator('#guide-progress').inner_text());self.assertIsNone(self.page.evaluate('emulator.device.snapshot(0).active'))
        self.clean()

if __name__=='__main__':unittest.main(verbosity=2)

File: tests/test_emulator_queue2_attacks.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {queue2Attacks} from '../tools/ui_emulator/web/queue2-attacks.mjs';
test('all S09-S12 variants have isolated guides',()=>{
 assert.deepEqual(queue2Attacks.slice(0,9).map(x=>x.id),['karma-menu','deauth','sae','handshake','blackout','global-handshaker','snifferdog','beacon','beacon-empty']);
 for(const entry of queue2Attacks){const g=entry.create();g.start({run:9});assert.equal(g.complete,false);assert.ok(g.story.steps.length>=3);g.accept({epoch:g.epoch-1,tab:0,run:10,running:true});assert.equal(g.step,0);g.setContext({tab:2,visible:true});g.accept({epoch:g.epoch,tab:2,run:10,running:true});assert.equal(g.step,0);g.leave();assert.equal(g.active,false);}
});

for(const id of ['deauth','sae','handshake','beacon'])test(id+' rejects stale, wrong target, hidden and changed run',()=>{
 const g=queue2Attacks.find(x=>x.id===id).create();g.start({run:4});
 const emit=s=>g.accept({tab:0,epoch:g.epoch,connected:true,...s});
 emit({scan:true,targets:['net-2'],beaconList:false});assert.equal(g.step,0);
 emit({scan:true,targets:['net-1'],beaconList:true,ssidCount:2});assert.equal(g.step,1);
 g.setContext({tab:0,visible:false});emit({popup:true,running:true,run:5,jobTargets:['net-1']});assert.equal(g.step,1);
 g.setContext({tab:0,visible:true});emit({popup:true,running:true,run:4,jobTargets:['net-1']});assert.equal(g.step,1);
 emit({popup:true,running:true,run:5,jobTargets:['net-1'],packets:24});assert.equal(g.step,2);
 emit({popup:true,running:true,run:6,packets:100,hsSuccess:true});assert.equal(g.step,2);
 emit({popup:true,running:true,run:5,packets:24,hsSuccess:true});assert.equal(g.step,2);
 emit({popup:true,running:true,run:5,packets:48,hsSuccess:true});assert.equal(g.step,3);
 emit({run:5,state:'cancelled',popup:false});assert.equal(g.complete,false);
 emit({run:5,state:'completed',popup:false,active:null});assert.equal(g.complete,true);
 g.start({run:5});assert.equal(g.step,0);assert.equal(g.complete,false);
});
for(const id of ['blackout','global-handshaker','snifferdog'])test(id+' requires confirmation cancellation before start',()=>{
 const g=queue2Attacks.find(x=>x.id===id).create();g.start({});const emit=s=>g.accept({tab:0,epoch:g.epoch,connected:true,...s});
 emit({running:true,run:1});assert.equal(g.step,0);emit({confirm:true});assert.equal(g.step,1);
 emit({confirm:false,popup:false,active:1});assert.equal(g.step,1);emit({confirm:false,popup:false,active:null});assert.equal(g.step,2);
 emit({running:true,run:2,packets:0,popup:true});assert.equal(g.step,3);emit({running:true,run:2,popup:true,packets:24,handshakes:1});assert.equal(g.step,4);
 emit({run:2,state:'completed',popup:false,active:null});assert.equal(g.complete,true);
});
import {readFile} from 'node:fs/promises';
import {SimulatedDevice} from '../tools/ui_emulator/model/device.mjs';
const seed=JSON.parse(await readFile(new URL('../docs/ui-emulator/neon-district.seed.json',import.meta.url),'utf8'));
test('Handshake reserves its target and module and freezes on cancellation',()=>{
 const d=new SimulatedDevice(seed),n=seed.networks[1],id=d.attackStart('grove','handshake',{bssids:[n.bssid]});
 assert.deepEqual(d.job(id).options.networkIds,[n.id]);assert.throws(()=>d.attackStart('grove','karma'),/busy/);
 d.advance(3000);assert.ok(d.job(id).result.packets>0);assert.ok(d.job(id).result.handshakes.every(x=>x.networkId===n.id));
 d.setModule('grove',{connected:false});assert.equal(d.job(id).state,'cancelled');const final=d.job(id);d.advance(6000);assert.deepEqual(d.job(id),final);
 assert.equal(d.snapshot('mbus').active,null);
});
for(const id of ['karma-menu'])test(id+' requires correct entry and a second fresh portal session',()=>{
 const g=queue2Attacks.find(e=>e.id===id).create();g.start({run:8});const emit=s=>g.accept({tab:0,epoch:g.epoch,connected:true,...s});
 emit(id==='karma-menu'?{scan:true}:{home:true});assert.equal(g.step,0);
 emit(id==='karma-menu'?{home:true}:{scan:true});assert.equal(g.step,1);emit({karma:true});assert.equal(g.step,2);
 emit({karma:true,running:true,run:9,packets:0});emit({karma:true,running:true,run:9,packets:24});assert.equal(g.step,4);
 emit({run:9,state:'completed',karmaConfig:true});assert.equal(g.step,5);
 emit({run:9,running:true,karmaPopup:true,ssids:['NEON-BAZAAR']});assert.equal(g.step,5);
 emit({run:10,running:true,karmaPopup:true,ssids:['wrong']});assert.equal(g.step,5);
 emit({run:10,running:true,karmaPopup:true,ssids:['NEON-BAZAAR'],packets:0});assert.equal(g.step,6);
 emit({run:10,running:true,karmaPopup:true,clients:1,packets:24});assert.equal(g.step,7);
 emit({run:10,state:'completed',popup:false,active:null});assert.equal(g.step,8);emit({home:true,karma:false});assert.equal(g.complete,true);
});
test('Beacon empty refusal needs an empty native list and no reserved operation',()=>{
 const g=queue2Attacks.find(e=>e.id==='beacon-empty').create();g.start({});const emit=s=>g.accept({tab:0,epoch:g.epoch,...s});
 emit({beaconList:true,ssidCount:2});emit({beaconList:true,ssidCount:0});assert.equal(g.step,2);
 emit({beacon:true,ssidCount:0,emptyRefusal:true,active:17});assert.equal(g.step,2);
 emit({beacon:true,ssidCount:0,emptyRefusal:true,active:null});assert.equal(g.step,3);
 emit({beacon:false,beaconList:false,active:null});assert.equal(g.complete,true);
});
